File: BOJ/1992.cpp
#include <iostream>
#include <string>

using namespace std;

int map[65][65];

void divide(int row, int col, int N);
int main(){
    // 첫째 줄에는 영상의 크기를 나타내는 숫자 N(1 ≤ N ≤ 64)
    int N;
    cin >> N;
    for (int i=0; i<N; i++){
        string s;
        cin >> s;
        for (int j=0; j<N; j++){
            map[i][j]=s[j]-'0';
        }
    }
    divide(0, 0, N);
    return 0;
}

void divide(int row, int col, int N){
    int start=map[row][col];
    for (int i=row; i<row+N; i++){
        for (int j=col; j<col+N; j++){
            // 다르면 N/2 사이즈로 쪼개기
            if(start!=map[i][j]){
                cout << "(";
                // 이차원 배열
                    // arr[0][0] arr[0][1] arr[0][2]
                    // arr[1][0] arr[1][1] arr[1][2]
                // 현재 위치 왼쪽 위 (row, col)
                divide(row, col, N/2);
                // 오른쪽 위
				divide(row, col+N/2, N/2);
                // 왼쪽 아래
				divide(row+N/2, col, N/2);
                // 오른쪽 아래
				divide(row+N/2, col+N/2, N/2);
				cout << ")";
				return;
            }
        }
    }
    cout << map[row][col];
}

File: BOJ JAVA/1992.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    public static StringBuilder sb = new StringBuilder();
    public static int[][] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        arr = new int[n][n];

        for (int i = 0; i < n; i++) {
            String str = br.readLine();
            for (int j = 0; j < n; j++) {
                arr[i][j] = str.charAt(j)-'0';
            }
        }

        br.close();

        divide(0, 0, n);

        System.out.println(sb);
    }

    // 주어진 범위가 모두 같은지 확인
    public static boolean check(int row, int col, int range){
        int num = arr[row][col];
        for (int i = row; i < row+range; i++) {
            for (int j = col; j < col+range; j++) {
                if(arr[i][j]!=num) return false;
            }
        }
        return true;
    }

    public static void divide(int row, int col, int range){
        if(check(row, col, range)){
            sb.append(arr[row][col]);
            return;
        }
        
        sb.append('(');
        for (int i = row; i < row+range; i+=range/2) {
            for (int j = col; j < col+range; j+=range/2) {
                divide(i, j, range/2);
            }
        }
        sb.append(')');
    }
}
